Removed reflection from composite read and write

YohDeadfall opened the pull request:
Fixes #2377.

roji commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberInfoOfStruct`2.cs:
```diff
@@ -0,0 +1,91 @@
+            _handler = handler;
+        }
+
+        public override async ValueTask Read(object composite, NpgsqlReadBuffer buffer, bool async)
```
At least in theory it's possible to accept a `ref TComposite composite`, avoiding the boxing inside ByReference. You can't have `ref` in async methods, but you *could* call all the async methods (e.g. `buffer.Ensure()`) and check whether the Tasks returned completed synchronously. If all the tasks completed synchronously (which they will in most cases) - great, we don't need to box.

But this would get tricky - if we can't complete synchronously we must box (copy the given ref TComposite into an object in the heap), and then we must return that object from the method. The API would look pretty weird but it could work... Also, this would probably break the abstraction that currently exists over struct and struct, although that doesn't seem very important.

What do you think? Am I missing something?

YohDeadfall replied:
Thanks for the idea, will try it.

YohDeadfall replied:
I remember why I didn't go this way. As you said, it won't look nice in that case. Let's do some boxing now? (: If we decide that it hurts performance to much, we can change the code.

By the way, I can optimize this code by making to `Read` methods which will accept `TComposite` and `ByReference<TComposite>` respectively. This change will eliminate one cast in the `Set` method. Don't know is it worth enough.

roji replied:
Is it really that ugly/complicated? On the one hand, I think the vast majority of cases would return synchronously, and so benefit from the elimination of the allocation. On the other hand, it's true that composite use in general isn't that widespread, and mapping structs to composites is probably even rarer.

If you really feel like this will make everything too ugly, I'm OK with leaving it out for now. But I might want to take a look myself to see what that involves etc...

YohDeadfall replied:
It's possible only for the sync way, but for other it's impossible becuase nobody know how fields are long. I have an idea, but will try it later.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/ByReference`1.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using System.Diagnostics;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
```
Yes, I'm planning to do it for all other type handlers too, but only after dropping unmapped composites and enums.

YohDeadfall commented on src/Npgsql/TypeHandling/NullableHandler.cs:
```diff
@@ -6,8 +6,8 @@
 {
-    delegate T ReadDelegate<T>(NpgsqlReadBuffer buffer, int columnLength, FieldDescription fieldDescription);
-    delegate ValueTask<T> ReadAsyncDelegate<T>(NpgsqlReadBuffer columnLength, int columnLen, bool async, FieldDescription fieldDescription);
+    delegate T ReadDelegate<T>(NpgsqlTypeHandler handler, NpgsqlReadBuffer buffer, int columnLength, FieldDescription fieldDescription);
```
There is a little value in it because of another refactoring. But if you want, I can make it optional.

roji replied:
Up to you - if you have something in mind which would obviate it, we can skip it. But at least for now it would make the code slightly better.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/ByReference`1.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using System.Diagnostics;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
```
Makes sense. By the way, I checked the Core FX repo and they prefer to omit back ticks too.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+                ? typeof(CompositeMemberInfoOfStruct<,>)
+                : typeof(CompositeMemberInfoOfClass<,>);
+
+            foreach (var compositeMember in compositeType.GetMembers(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
```
I have the same concerns about that too, but for compatibility did it this way. If you and @austindrenski are okay with dropping non-public member support, will do it.

austindrenski replied:
I lean toward not including non-public members by default. It seems both simpler and more appropriate to just respect the CLR visibility, barring explicit info from the user that "yes, I _do_ want to map this specific non-public member".

If we want to support non-public members, maybe we can look into (and try to match) the EF Core decisions on non-public mapping?

YohDeadfall replied:
Okay, will drop in a new PR to track that in issues and release notes. Otherwise, there is a change that we will forget about that and won't notice our users. Opened #2405 to track progress.

roji replied:
I took a look to confirm and [we currently look at public members only](https://github.com/npgsql/npgsql/blob/94b16acd54c805b74226a002a4f2f3816020b05f/src/Npgsql/TypeHandlers/UnmappedCompositeHandler.cs#L224). This wouldn't be a breaking change.

YohDeadfall replied:
Yes, fixed a test.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+                Type compositeMemberType;
+                switch (compositeMember)
+                {
+                    case FieldInfo compositeFields:
```
By the way, should we support fields? I implemented it for compatibility, but not sure that it's the right way.

/cc @austindrenski

roji replied:
I don't see any particular reason to drop support for fields - especially for structs it seems somewhat common to just have public fields... Do you see a problem here?

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+                    continue;
+
+                if (mappedMembers[index] != null)
+                    throw new InvalidOperationException("");
```
Oops, forgot to add it (: This will happen if several CLR members are mapped to the same field. It's possible only with `PgNameAttribute`.

YohDeadfall replied:
Changed the exception type and added a message.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+            var unmappedFields = string.Join(", ", mappedMembers
+                .Where(member => member == null)
+                .Select((member, memberIndex) => mappedFields[memberIndex].Name));
+            throw new InvalidOperationException($"PostgreSQL composite type {postgresType.DisplayName} contains fields {unmappedFields} which could not match any on CLR type {compositeType.Name}");
```
Maybe each field?

roji replied:
Could be. Another option is to just print the first field - this would prevent the exception message from getting too long.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+                .Select((member, memberIndex) => mappedFields[memberIndex].Name));
+            throw new InvalidOperationException($"PostgreSQL composite type {postgresType.DisplayName} contains fields {unmappedFields} which could not match any on CLR type {compositeType.Name}");
+
+            PostgresCompositeType.Field GetField(string name, out int index)
```
Will do.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/ByReference`1.cs:
```diff
@@ -0,0 +1,17 @@
+            Value = value;
+        }
+
+        public static readonly bool IsTrue = typeof(T).IsValueType;
```
Had no better name in mind that that, and yes, it's confusing (:

roji replied:
It would be good to check that there's an actual perf benefit to having this - maybe just putting `typeof(T).IsValueType` everywhere is already optimized by the JIT (if it's not it should be :)).

Otherwise I'd just call the property `IsValueType`...

YohDeadfall replied:
By the way, there is `RuntimeHelpers.IsReferenceOrContainsReferences<T>` which isn't available for all out targets.

roji replied:
Re `RuntimeHelpers.IsReferenceOrContainsReferences<T>`, if this returns true on something that contains a reference field, then it wouldn't be good for us... A struct which contains a string is still a struct and should be treated as such.

YohDeadfall replied:
Checked, you're right. The previous version (`IsReference<T>`) was designed to say the same thing as `Type.IsValue`. Unfortunately, there is no documentation about it.

YohDeadfall commented:
There are no benchmarks because I had no time, but am planning to write them and tests too.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+
+        public static MappedCompositeHandler<T> Create(PostgresCompositeType postgresType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        {
+            var compositeType = typeof(T);
```
It's a performance optimization to get the type once. The method shouldn't be called often, but why not to optimize the code if it's possible. Would it be better to add `clr` prefix for members and the type local, and `postgres` for fields?

roji replied:
This code runs only once when the physical connection is created (or when typed are reloaded), so really really not perf-sensitive...

But if you really feel strongly about it, then maybe `clr` and `pg` prefixes could help make sure there's no confusion (`postgres` is loooong).

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberInfoOfClass`2.cs:
```diff
@@ -0,0 +1,84 @@
+
+            var setMethod = propertyInfo.GetSetMethod();
+            if (setMethod?.IsPublic == true)
+                _set = (SetMember)Delegate.CreateDelegate(typeof(SetMember), setMethod);
```
To avoid boxing and dynamic binding. So I use delegates to achieve the same behavior as calling a getter/setter directly.

austindrenski replied:
Interesting... I'd always just assumed that delegate creation just pushed the boxing out of sight (but still incurred it when called). Thanks for the info!

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberInfoOfClass`2.cs:
```diff
@@ -0,0 +1,84 @@
+                _get = (GetMember)Delegate.CreateDelegate(typeof(GetMember), getMethod);
+
+            var setMethod = propertyInfo.GetSetMethod();
+            if (setMethod?.IsPublic == true)
```
You are right, will fix.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberInfoOfStruct`2.cs:
```diff
@@ -0,0 +1,91 @@
+            ? NullableHandler<TMember>.ValidateAndGetLength(_handler, _get(ref composite), ref lengthCache, null)
+            : _handler.ValidateAndGetLength(_get(ref composite), ref lengthCache, null);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Nope, you cannot mix `ref` with `async` now, but it's a compiler limitation. There is an issue to fix that behavior and allow `ref` usage between `await`'s, but I don't have a link to it (:

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/IMappedCompositeTypeHandlerFactory.cs:
```diff
@@ -0,0 +1,14 @@
+    public interface IMappedCompositeTypeHandlerFactory
+    {
+        /// <summary>
+        /// The name translator used for this enum.
```
You found an old bug in out documentation (: Will fix it too.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+            if (fieldCount != _members.Length)
+                throw new InvalidOperationException($"pg_attributes contains {_members.Length} fields for type {PgDisplayName}, but {fieldCount} fields were received.");
+
+            var composite = new T();
```
Yep, will do, but in a separate PR.

roji replied:
Re constructor mapping, let's not do this right away - nobody has ever asked for it and there's lots of other work :) We can open an issue so that people can vote though.

IMHO, Npgsql's composite doesn't pretend to be a full-fledged mapping technology like EF Core, and composite types are unlikely to be used by users in the way EF Core entities are used. So I'd rather hear some user requests for this (and understand the usecase) before jumping too quickly with it...

YohDeadfall replied:
> Re constructor mapping, let's not do this right away - nobody has ever asked for it and there's lots of other work :) We can open an issue so that people can vote though.

I remember I saw an issue about it from F# community. Anyway don't forget about immutability, for structures it's only one good way as you know.

roji replied:
Yeah. I think it could make sense and I'm not against it, just saying we should wait until this feature is requested, that's all (there's lots of stuff to do).

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+
+            var unmappedFields = string.Join(", ", mappedMembers
+                .Where(member => member == null)
+                .Select((member, memberIndex) => mappedFields[memberIndex].Name));
```
It's a bug. Wil fix.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/UnmappedCompositeTypeHandlerFactory.cs:
```diff
@@ -0,0 +1,17 @@
+        internal UnmappedCompositeTypeHandlerFactory(INpgsqlNameTranslator nameTranslator)
+        {
+            _nameTranslator = nameTranslator;
+        }
```
Let's leave it as is, because the next step is to drop it. See #2403.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs:
```diff
@@ -0,0 +1,152 @@
+            foreach (var compositeMember in compositeType.GetMembers(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
+            {
+                Type compositeMemberType;
+                switch (compositeMember)
```
I wrote the code using an old Visual Studio where is no C# 8.0. Am planning to reset my VM this weekend and instal the latest one, so will be able to use new features.

YohDeadfall replied:
Leaving as it is because of `continue`. A null check looks excessive.

austindrenski commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,150 @@
+                for (pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                {
+                    pgField = pgFields[pgFieldIndex];
+                    if (pgField.Name == name) break;
```
_This isn't my area of expertise, so please forgive what may actually be a silly question:_

Is there a perf reason to avoid checking `pgFields[pgFieldIndex].Name == name` directly? Is repeatedly reassigning the local cheaper than the one additional index access?

For example, assuming that we're dealing with a composite of 4 fields, `(a, b, c, d)` and we want to get field `a`:

The current version (`pgField.Name == name`):
- indexes `pgFields` 4 times
- accesses `.Name` 4 times
- writes `pgField` 4 times

But if it checked `pgFields[pgFieldIndex].Name == name` directly, it would need to:
- indexes `pgFields` 5 times
- accesses `.Name` 4 times 
- writes `pgField` 1 time.

YohDeadfall replied:
Don't know and will check.

YohDeadfall replied:
Checked and verified that a local variable could improve perf when an iteration goes over an array. Probably it's the right behavior because there could be side effects and the JIT takes that into account. The last two tests (without direct return) demonstrate the current scenario when the result is used after a cycle.

One small note before you will see the test results. This isn't applicable to the commented code since `pgFields` isn't an array but an `IReadOnlyList<T>`. Therefore, the JIT must call the indexer each time.

```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.16299.1029 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
Frequency=2062499 Hz, Resolution=484.8487 ns, Timer=TSC
.NET Core SDK=2.2.104
  [Host]     : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
```

|                                Method |     Mean |     Error |    StdDev |
|-------------------------------------- |---------:|----------:|----------:|
|      Iteration_WithLocal_DirectReturn | 1.972 ns | 0.1022 ns | 0.1177 ns |
|   Iteration_WithoutLocal_DirectReturn | 2.138 ns | 0.1012 ns | 0.1039 ns |
|    Iteration_WithLocal_NoDirectReturn | 1.713 ns | 0.0952 ns | 0.0978 ns |
| Iteration_WithoutLocal_NoDirectReturn | 1.956 ns | 0.1051 ns | 0.1438 ns |

```csharp
private readonly Field[] pgFields = new[]
{
    new Field("one"),
    new Field("two"),
    new Field("three"),
};

[Benchmark]
public Field Iteration_WithLocal_DirectReturn()
{
    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        var pgField = pgFields[pgFieldIndex];
        if (pgField.Name == "three")
        {
            return pgField;
        }
    }

    return null;
}

[Benchmark]
public Field Iteration_WithoutLocal_DirectReturn()
{
    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        if (pgFields[pgFieldIndex].Name == "three")
        {
            return pgFields[pgFieldIndex];
        }
    }

    return null;
}

[Benchmark]
public Field Iteration_WithLocal_NoDirectReturn()
{
    Field pgField = null;

    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        pgField = pgFields[pgFieldIndex];

        if (pgField.Name == "three")
        {
            break;
        }
    }

    return pgField;
}

[Benchmark]
public Field Iteration_WithoutLocal_NoDirectReturn()
{
    Field pgField = null;

    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        if (pgFields[pgFieldIndex].Name == "three")
        {
            pgField = pgFields[pgFieldIndex];
            break;
        }
    }

    return pgField;
}
```

austindrenski replied:
Very interesting, thanks for taking a look at it!

austindrenski commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeTypeHandlerFactory.cs:
```diff
@@ -0,0 +1,21 @@
+            => MappedCompositeHandler<T>.Create((PostgresCompositeType)pgType, conn.Connector.TypeMapper, NameTranslator);
+
+        protected override NpgsqlTypeHandler<T> Create(NpgsqlConnection conn)
+            => throw new InvalidOperationException($"Expect {nameof(PostgresType)}");
```
Should this be a `NotSupportedException`?

YohDeadfall replied:
Not mine. Just moved the old code to a separate file.

By the way, I think that this method can be removed from all factories, and the behavior should be changed to provide a Postgres type directly to the constructor. This means that the internal method must be `protected` not `internal`.

roji replied:
The point here was that handler in general don't care about the PostgresType at all - except for CompositeTypeHandler. At the same time, they need to reference it for the metadata APIs (e.g. `NpgsqlDataReader.GetDataTypeName()`.

So we could add a constructor to each and every handler as @YohDeadfall suggests, and pass it up to the base class's constructor, or we can have it the current way where a type handler doesn't need to care about the PostgreSQL type, unless they choose to (like the composite handler). I remember not being too sure what I preferred, by the way, but I ended up doing this.

The specific exception type doesn't really matter since the method is never supposed to get invoked.

austindrenski commented on src/Npgsql/TypeHandling/NullableHandler.cs:
```diff
@@ -43,21 +43,23 @@ static class NullableHandler
-            => handler.WriteWithLengthInternal(value.Value, buffer, lengthCache, parameter, async);
+            => value.HasValue
+            ? handler.WriteWithLengthInternal(value.Value, buffer, lengthCache, parameter, async)
+            : handler.WriteWithLengthInternal(DBNull.Value, buffer, lengthCache, parameter, async);
```
Also, perhaps inline with:

```c#
WriteWithLengthInternal(value.HasValue ? value.Value : DBNull.Value, buffer, lengthCache, parameter, async);
```

YohDeadfall replied:
`WriteWithLengthInternal` is generic. So if I follow you suggestion then there will be boxing and handler lookup inside the handler.

austindrenski replied:
Ope, didn't realize that, please disregard.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,150 @@
+
+                return reference.Value;
+            }
+            else
```
Yes, but I would like to leave it to split the cases for visibility because the JIT will compile only one of them into machine code.

austindrenski replied:
Fair enough.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,150 @@
+                var attr = clrMember.GetCustomAttribute<PgNameAttribute>();
+                var name = attr != null
+                    ? attr.PgName
+                    : nameTranslator.TranslateMemberName(clrMember.Name);
```
Unfortunately, these expressions aren't equal to each other. `PgNameAttribute` doesn't check that its argument. Therefore, the name can be `null`. I think in that case we should fail with a field not mapped exception.

Probably the behavior should be changed, but let's do it as a separate pull request. Otherwise, there is a change that I will refactor the whole driver here (: Don't resolve this comment, I'll open an issue to track that.

austindrenski replied:
> Unfortunately, these expressions aren't equal to each other.

You are correct, but I would still opt to treat them the same anyways. For one thing, it's [what we do in the EF Core provider](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/6b58e6c280d15318aebbd8e259bb7c532cccfdd1/src/EFCore.PG/Storage/Internal/Mapping/NpgsqlEnumTypeMapping.cs#L61).

> I think in that case we should fail with a field not mapped exception.

That seems a little too aggressive to me. If it's a typo (e.g. forgot to add the name) and the CLR name isn't going to map, we'll fail soon anyways. So why not allow for some flexibility by falling back to the name translator at this point?

YohDeadfall replied:
Makes sense, will change.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,150 @@
+                for (pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                {
+                    pgField = pgFields[pgFieldIndex];
+                    if (pgField.Name == name) break;
```
Do you realy want it? The expression fits 95 characters (:

austindrenski replied:
Heh, well I would for consistency, but it's fine either way.

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,150 @@
+                ? typeof(CompositeMemberHandlerOfStruct<,>)
+                : typeof(CompositeMemberHandlerOfClass<,>);
+
+            foreach (var clrMember in clrType.GetMembers(BindingFlags.Instance | BindingFlags.Public))
```
This one for user code. We are getting members which can bee mapped here. On line 135 I specify visibility for our member handler. These are different cases.

austindrenski replied:
Ah, makes sense!

austindrenski commented on src/Npgsql/TypeHandlers/CompositeHandlers/IsValueType.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    static class IsValueType<T>
+    {
+        public static readonly bool Value = typeof(T).IsValueType;
```
@roji [wrote](https://github.com/npgsql/npgsql/pull/2402#discussion_r270114598):

> It would be good to check that there's an actual perf benefit to having this - maybe just putting `typeof(T).IsValueType` everywhere is already optimized by the JIT (if it's not it should be :)).
>
> Otherwise I'd just call the property `IsValueType`...

FYI tracked down the [discussion on intrinsic support](https://github.com/dotnet/coreclr/issues/23208).

roji replied:
Thanks for tracking this down @austindrenski! At least it's on their radar...

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/IDynamicCompositeTypeHandlerFactory.cs:
```diff
@@ -0,0 +1,6 @@
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+#pragma warning disable CA1040    // Avoid empty interfaces
+    interface IDynamicCompositeTypeHandlerFactory { }
```
Deleted.

roji commented on src/Npgsql/TypeHandlers/CompositeHandlers/ByReference.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    sealed class ByReference<T>
```
nit: constraint T to struct (mostly for documentation value), this would allow you to remove the `Debug.Assert()` down below.

YohDeadfall replied:
Not possible without return of `Unsafe` which was removed by @khellang. This type is used in `MappedCompositeHandler<T>` which doesn't have a constraint on `T`.

khellang replied:
I don't think I've ever touched this type? 🤔

roji replied:
Makes sense - at least the part where it's used by `MappedCompositeHandler<T>` which itself isn't constrained. It's not important.

YohDeadfall replied:
@khellang You have removed `Unsafe` via #2310 and #2313.

khellang replied:
Right, but not in this type. There are still valid use cases for `Unsafe` 😉

roji commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberHandlerOfClass.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    sealed class CompositeMemberHandlerOfClass<TComposite, TMember> : CompositeMemberHandler<TComposite>
```
Proposal: rename to CompositeClassMemberHandler and CompositeStructMemberHandler.

YohDeadfall commented:
```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Xeon CPU E5-2687W 0 3.10GHz, 2 CPU, 32 logical and 16 physical cores
.NET Core SDK=3.0.100-preview3-010431
  [Host]     : .NET Core 3.0.0-preview3-27503-5 (CoreCLR 4.6.27422.72, CoreFX 4.7.19.12807), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview3-27503-5 (CoreCLR 4.6.27422.72, CoreFX 4.7.19.12807), 64bit RyuJIT
```

### Mapped composites

| Method |                Value |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|   Read | Npgsq(...)Class [50] | 2.161 us | 0.0536 us | 0.1580 us | 2.184 us | 462,655.1 | 0.0515 |     - |     - |     328 B |
|  Write | Npgsq(...)Class [50] | 2.575 us | 0.0744 us | 0.2169 us | 2.641 us | 388,298.9 | 0.0916 |     - |     - |     576 B |

### Unmapped composites

| Method |                Value |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|   Read | Npgsq(...)Class [50] | 3.836 us | 0.0840 us | 0.2476 us | 3.862 us | 260,720.3 | 0.0839 |     - |     - |     544 B |
|  Write | Npgsq(...)Class [50] | 4.004 us | 0.2896 us | 0.8540 us | 4.272 us | 249,735.3 | 0.0992 |     - |     - |     624 B |

roji commented:
Really nice to see the perf improvement @YohDeadfall! Will try to review this and approve ASAP.

YohDeadfall commented:
Changed composite materialization to avoid `Activator.CreateInstance<T>` and added tests for struct handling. Ready for final review.

```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Xeon CPU E5-2687W 0 3.10GHz, 2 CPU, 32 logical and 16 physical cores
.NET Core SDK=3.0.100-preview3-010431
  [Host]     : .NET Core 3.0.0-preview3-27503-5 (CoreCLR 4.6.27422.72, CoreFX 4.7.19.12807), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview3-27503-5 (CoreCLR 4.6.27422.72, CoreFX 4.7.19.12807), 64bit RyuJIT
```

### Mapped class

| Method |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|   Read | 1.889 us | 0.1177 us | 0.3471 us | 2.063 us | 529,328.0 | 0.0496 |     - |     - |     328 B |
|  Write | 2.323 us | 0.1592 us | 0.4694 us | 2.535 us | 430,461.0 | 0.0916 |     - |     - |     576 B |

### Mapped struct

| Method |       Mean |    Error |    StdDev |     Median |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |-----------:|---------:|----------:|-----------:|------------:|-------:|------:|------:|----------:|
|   Read | 1,285.7 ns | 25.96 ns |  72.37 ns | 1,299.1 ns |   777,814.3 | 0.0305 |     - |     - |     192 B |
|  Write |   838.7 ns | 52.80 ns | 155.69 ns |   898.2 ns | 1,192,331.4 | 0.0458 |     - |     - |     288 B |

### Unmapped class

| Method |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|   Read | 3.962 us | 0.1122 us | 0.3308 us | 4.011 us | 252,423.5 | 0.0839 |     - |     - |     544 B |
|  Write | 3.524 us | 0.3166 us | 0.9334 us | 3.980 us | 283,806.9 | 0.0992 |     - |     - |     624 B |

### Unmapped struct

| Method |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|   Read | 3.134 us | 0.0627 us | 0.1377 us | 3.137 us | 319,079.4 | 0.0572 |     - |     - |     368 B |
|  Write | 2.390 us | 0.0741 us | 0.2184 us | 2.451 us | 418,380.1 | 0.0610 |     - |     - |     384 B |

roji commented on src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs:
```diff
@@ -0,0 +1,155 @@
+            }
+            else
+            {
+                var composite = s_constructor();
```
It's a good idea to eliminate instantiation with Activator, but... since T is already constrained to `new()`, why not just do `var composite = new T()`?

If we need to keep the code-generator for some reason, then a naming nit: we don't usually use naming such as `s_constructor`, how about `ClassConstructor` or `ClassInstantiator`?

YohDeadfall replied:
Because `new T()` actually is `Activator.CreateInstance<T>` because compile doesn't know an actual constructor which will be used. For structs it doesn't exist at all (:

Done, but i used the standard notation in .NET. I think about switching to it.

roji replied:
> Because new T() actually is Activator.CreateInstance<T> because compile doesn't know an actual constructor which will be used.

I didn't know that - but I can see that you're right on sharplabs.io. Do you know of any blog/article that explains what it's like this? I'd have thought that as long as there's a `new()` type constraint, than an efficient `new T()` should be possible, without using Activator.

khellang replied:
See https://stackoverflow.com/a/367648/682105

roji replied:
Wow, thanks for all those links @khellang :) Makes sense. Added a proposal on SO to have the JIT recognize `Activator.CreateInstance<T>()` and emit the right thing.

YohDeadfall replied:
A proposal on SO? By the way, I'm planning to open proposal in the Core FX repo in a few days and provide a non-intrinsic implementation.

roji replied:
Maybe proposal was the wrong word, more like a thought :)

In any case, [see this comment](https://github.com/dotnet/coreclr/issues/7592#issuecomment-481699119)...

YohDeadfall commented on src/Npgsql/TypeHandlers/CompositeHandlers/CompositeMemberHandlerOfClass.cs:
```diff
@@ -0,0 +1,94 @@
+                ? await NullableHandler<TMember>.ReadAsync(_handler, buffer, length, async)
+                : await _handler.Read<TMember>(buffer, length, async);
+
+            _set((TComposite)composite, value);
```
Done.

roji commented:
Feel free to merge @YohDeadfall (don't forget to squash all those commits...).

YohDeadfall commented:
Will merge when answer on https://github.com/npgsql/npgsql/pull/2402#discussion_r270638544.

roji commented:
:champagne:
